### ISSUE TITLE
Error applying view update to 172.30.0.185: exceptions::mutation_write_timeout_exception

### ISSUE THREAD

**Issue by yarongilor** [scylladb/scylla#4900]
Installation details
Scylla version (or git commit hash): 3.1.0.rc3-0.20190816.d06bcef3b
Cluster size: 6
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-07f2007dc543eced5

scenario: 50GB longevity on starting nemesis of destroy-data-then-repair
error:
```
< t:2019-08-21 19:14:54,198 f:cluster.py      l:1212 c:sdcm.cluster         p:DEBUG > 2019-08-21T19:11:39+00:00  ip-172-30-0-228 !ERR     | scylla.bin: [shard 4] view - Error applying view update to 172.30.0.185: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_first_name - received only 0 responses from 1 CL=ANY.)

```

logs:
```
[jenkins@aws-scylla-qa-builder latest]$ grep -i uploading ./sct.log 
< t:2019-08-24 03:22:21,843 f:common.py       l:328  c:utils                p:INFO  > Uploading '/sct/20190819-131620-166949/longevity-tls-50gb-4d-3-1-monitor-set-327565b4/longevity-tls-50gb-4d-3-1-monitor-node-327565b4-1/grafana-screenshot-per-server-metrics-nemesis-20190824_032210-0.png' to https://cloudius-jenkins-test.s3.amazonaws.com/327565b4-2837-45a5-9c5a-e02d7c6c0624/grafana-screenshot-per-server-metrics-nemesis-20190824_032210-0.png
< t:2019-08-24 03:23:34,656 f:common.py       l:328  c:utils                p:INFO  > Uploading '/sct/20190819-131620-166949/longevity-tls-50gb-4d-3-1-monitor-set-327565b4/longevity-tls-50gb-4d-3-1-monitor-node-327565b4-1/grafana-screenshot-overview-metrics-20190824_032323-0.png' to https://cloudius-jenkins-test.s3.amazonaws.com/327565b4-2837-45a5-9c5a-e02d7c6c0624/grafana-screenshot-overview-metrics-20190824_032323-0.png
< t:2019-08-24 03:24:35,961 f:cluster.py      l:3740 c:sdcm.cluster         p:INFO  > Uploading 'annotations.json' to https://cloudius-jenkins-test.s3.amazonaws.com/327565b4-2837-45a5-9c5a-e02d7c6c0624/annotations.json
< t:2019-08-24 03:24:39,222 f:common.py       l:328  c:utils                p:INFO  > Uploading '/sct/events_log.zip' to https://cloudius-jenkins-test.s3.amazonaws.com/327565b4-2837-45a5-9c5a-e02d7c6c0624/events_log.zip
< t:2019-08-24 03:25:56,959 f:common.py       l:328  c:utils                p:INFO  > Uploading '/sct/20190819-131620-166949/job_log.zip' to https://cloudius-jenkins-test.s3.amazonaws.com/327565b4-2837-45a5-9c5a-e02d7c6c0624/job_log.zip
```

**Comment by yarongilor** [comment]
Uploaded logs to gcloud in:
scratch.scylladb.com/issues/sct_4901.log.gz

**Comment by slivne** [comment]
@yarongilor do you have metrics

We had many stalls around this period - could that explain this

**Comment by psarna** [comment]
One question to start: in the logs I see several errors like this one: `WARNING | scylla.bin: [shard 0] stream_session - [Stream #51e9e660-c536-11e9-9d88-000000000000] stream_transfer_task: Fail to send to 172.30.0.228:0: std::runtime_error (Peer failed to pro
cess mutation_fragment peer=172.30.0.228, plan_id=51e9e660-c536-11e9-9d88-000000000000, cf_id=fbed93f0-c289-11e9-b393-000000000009)
`
It comes from the old streaming - why isn't 3.1 using row_level streaming for repairs? `git log` suggests that it's already in for 3.1, isn't it?

As for stalls, I remember seeing several patches that avoided them during streaming, but all were related with row level repair:
~~~
06c84be257e76b097b3a7ab6fe2e2e78715a26fc repair: Futurize get_repair_rows_size and row_buf_size
809c992b30abbe6207dbdbfdbcc6fe76bee92c82 repair: Avoid calling get_repair_rows_size in get_sync_boundary
4d41f8e57ed8c5585412afcbb9180e564c0d5852 repair: Futurize row_buf_csum
0ef167c9c8aa8fbcf578bbba5fe2f655687ca2ec repair: Yield inside get_set_diff
~~~

**Comment by denesb** [comment]
> It comes from the old streaming - why isn't 3.1 using row_level streaming for repairs? git log suggests that it's already in for 3.1, isn't it?

Repair doesn't use streaming, it uses its own thing. This *is* the new streaming on top of rpc streaming.

**Comment by psarna** [comment]
Well, then these four commits are not applied to 3.1, so they can as well be the source of the stalls. And these stalls seem to cause broken semaphore errors too.

Also, I see that there are many `no such keyspace` exceptions, is it a known issue?
`std::_Nested_exception<no_such_keyspace> (Can't find a keyspace keyspace_snappy): std::out_of_range (_Map_base::at)`, `Failed to apply mutation from 172.30.0.154#0: std::_Nested_exception<schema_version_loading_failed> (Failed to load schema version c7d0c698-37fe-3a92-b2d8-ddc3f4619bd1): std::_Nested_exception<no_such_keyspace> (Can't find a keyspace keyspace_deflate): std::out_of_range (_Map_base::at)`, etc.

**Comment by slivne** [comment]
@psarna - the stalls you are seeing in repair - are they matching the fixes from asias - we do not backport "blindly" at this point. If there are stalls in the log - can you please "prove" that they "will" be fixed by these patches all/some.

**Comment by psarna** [comment]
@slivne it was just my hypothesis, the logs give little context aside from the fact that the stalls occurred while streaming was happening. I can try to reproduce and compare if the stalls appear without the patches and disappear after applying them.

**Comment by slivne** [comment]
On Mon, Sep 9, 2019 at 10:26 AM Piotr Sarna <notifications@github.com>
wrote:

> @slivne <https://github.com/slivne> it was just my hypothesis, the logs
> give little context aside from the fact that the stalls occurred while
> streaming was happening. I can try to reproduce and compare if the stalls
> appear without the patches and disappear after applying them.
>

extract the backtrace from the stalls  - one comment (avi's backported some
patches for the stall detector - it will be part of rc6 - so maybe it would
be better to use that - instead of working on stalls that are misleading)

**Comment by psarna** [comment]
@slivne Stalls in the provided logs don't have their backtraces logged - if they did, that's what I would have started with. I guess the `print_with_backtrace` call in stall detector failed for some reason?

**Comment by slivne** [comment]
@psarna ami for rc6 is available please take it and run some tests in parallel to QA trying to reproduce via SCT

**Comment by slivne** [comment]
Since reproducing this seems hard and releasing a version that has stalls that may break connection is also bad - we will do the backports from https://github.com/scylladb/scylla/issues/4900#issuecomment-528745982 and include them and test

**Comment by slivne** [comment]
@asias can you try and force stalls on 3.1 with repair - we need to see if we need to backport the fixes you have in master for row level repair stalls

**Comment by asias** [comment]
The stall fixes are for 3.2 only because 3.1 uses very small row buffer while 3.2 uses big row buffer which can cause stall. 

@psarna @denesb since you looked at this issue in the past. Could you share the backtrace of the stall you saw in this issue?

**Comment by asias** [comment]
> Uploaded logs to gcloud in:
> scratch.scylladb.com/issues/sct_4901.log.gz

Are you sure the log for this issue is scratch.scylladb.com/issues/sct_4901.log.gz not scratch.scylladb.com/issues/sct_4900.log.gz?

**Comment by asias** [comment]
> Well, then these four commits are not applied to 3.1, so they can as well be the source of the stalls. And these stalls seem to cause broken semaphore errors too.

What backtrace of the stall did you see. The stall fix patches are for 3.2 only (after we switch to use very big repair row buffer). How long is the mutation_write_timeout_exception? 3.1 is not supposed to have huge stalls even without the stall fix patches.

**Comment by yarongilor** [comment]
> > Uploaded logs to gcloud in:
> > scratch.scylladb.com/issues/sct_4901.log.gz
> 
> Are you sure the log for this issue is scratch.scylladb.com/issues/sct_4901.log.gz not scratch.scylladb.com/issues/sct_4900.log.gz?

i am positive about that, thanks.

**Comment by asias** [comment]
> > > Uploaded logs to gcloud in:
> > > scratch.scylladb.com/issues/sct_4901.log.gz
> > 
> > 
> > Are you sure the log for this issue is scratch.scylladb.com/issues/sct_4901.log.gz not scratch.scylladb.com/issues/sct_4900.log.gz?
> 
> i am positive about that, thanks.

In another issue, I saw you uploaded to this same url:
https://github.com/scylladb/scylla/issues/4901#issuecomment-526925984
```
Uploaded logs to gcloud in:
scratch.scylladb.com/issues/sct_4901.log.gz
```

**Comment by yarongilor** [comment]
> > > > Uploaded logs to gcloud in:
> > > > scratch.scylladb.com/issues/sct_4901.log.gz
> > > 
> > > 
> > > Are you sure the log for this issue is scratch.scylladb.com/issues/sct_4901.log.gz not scratch.scylladb.com/issues/sct_4900.log.gz?
> > 
> > 
> > i am positive about that, thanks.
> 
> In another issue, I saw you uploaded to this same url:
> [#4901 (comment)](https://github.com/scylladb/scylla/issues/4901#issuecomment-526925984)
> 
> ```
> Uploaded logs to gcloud in:
> scratch.scylladb.com/issues/sct_4901.log.gz
> ```

That is correct. the 2 issues are reported in one log file.

**Comment by slivne** [comment]
@asias any luck in looking at the logs ? are the stalls related

**Comment by asias** [comment]
> @slivne Stalls in the provided logs don't have their backtraces logged - if they did, that's what I would have started with. I guess the `print_with_backtrace` call in stall detector failed for some reason?

Hello @psarna, so you failed to extract any of the backtraces for the stalls with the log provided?

**Comment by asias** [comment]
> @asias any luck in looking at the logs ? are the stalls related

I have seen the logs but I can not to decode the backtrace in the log because I can not find the rpm for the version 3.1.0.rc3-0.20190816.d06bcef3b on our download repo.

**Comment by bhalevy** [comment]
> > @asias any luck in looking at the logs ? are the stalls related
> 
> I have seen the logs but I can not to decode the backtrace in the log because I can not find the rpm for the version 3.1.0.rc3-0.20190816.d06bcef3b on our download repo.

@hagitsegev please help @asias to locate the rpm.
Since this is 3.1.0.rc3 we expect addr2line to be helpful, but you may hit problems with gdb in this version and the relocatable package.

**Comment by slivne** [comment]
@hagitsegev ping

**Comment by hagitsegev** [comment]
Sorry for the delay. For some reason I missed the first request of 4 days ago.
I didn't hit problems on rc3. Jenkins centos-rpm build was number 90, but it does not exist anymore.
What is "gdb"?

**Comment by asias** [comment]
> Sorry for the delay. For some reason I missed the first request of 4 days ago.
> I didn't hit problems on rc3. Jenkins centos-rpm build was number 90, but it does not exist anymore.
> What is "gdb"?

I am not sure what problems are you referring to. What I am looking for is the scylla rpm for the version 3.1.0.rc3-0.20190816.d06bcef3b which was used to run the test. I couldn't find it on the download page. Can you still find it? If it does not exist any more, it's fine. We can move forward and ignore the core captured here.

Since we are waiting for scylla 3.1 with commit 834b92b to reproduce "Peer failed to process mutation_fragment after repair #4901" anyway. If it we have lots of stall, hopefully it will reproduce.

**Comment by slivne** [comment]
@yarongilor / @roydahan are we getting this stalls or errors in 3.1.rc7 if so can you please post the info on this issue so @asias can take a look

**Comment by tgrabiec** [comment]
On Fri, Sep 6, 2019 at 1:22 PM Piotr Sarna <notifications@github.com> wrote:

> Well, then these four commits are not applied to 3.1, so they can as well
> be the source of the stalls. And these stalls seem to cause broken
> semaphore errors too.
>
> Also, I see that there are many no such keyspace exceptions, is it a
> known issue?
>
There was https://github.com/scylladb/scylla/issues/5011 which could cause
this.

**Comment by yarongilor** [comment]
found a reproduction of the issue in 1TB longevity log.
ami-0f02b94adb41f1f9c
Scylla version 3.1.0.rc6-0.20190908.2ec036f50
```
/jenkins/slave/workspace/scylla-3.1/longevity/longevity-1tb-7days/scylla-cluster-tests/20190911-122051-005945/sct.log:< t:2019-09-15 06:06:37,644 f:cluster.py      l:1218 c:sdcm.cluster         p:DEBUG > 2019-09-15T02:09:26+00:00  ip-10-0-49-55 !ERR     | scylla: [shard 10] view - Error applying view update to 10.0.141.40: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_first_name - received only 0 responses from 1 CL=ANY.)
/jenkins/slave/workspace/scylla-3.1/longevity/longevity-1tb-7days/scylla-cluster-tests/20190911-122051-005945/sct.log:< t:2019-09-15 06:06:38,733 f:cluster.py      l:1218 c:sdcm.cluster         p:DEBUG > 2019-09-15T02:09:26+00:00  ip-10-0-49-55 !ERR     | scylla: [shard 5] view - Error applying view update to 10.0.141.40: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_last_name - received only 0 responses from 1 CL=ANY.)
/jenkins/slave/workspace/scylla-3.1/longevity/longevity-1tb-7days/scylla-cluster-tests/20190911-122051-005945/sct.log:< t:2019-09-15 06:06:39,692 f:cluster.py      l:1218 c:sdcm.cluster         p:DEBUG > 2019-09-15T02:09:26+00:00  ip-10-0-49-55 !ERR     | scylla: [shard 3] view - Error applying view update to 10.0.141.40: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_first_name - received only 0 responses from 1 CL=ANY.)
/jenkins/slave/workspace/scylla-3.1/longevity/longevity-1tb-7days/scylla-cluster-tests/20190911-122051-005945/sct.log:< t:2019-09-15 06:06:40,398 f:cluster.py      l:1218 c:sdcm.cluster         p:DEBUG > 2019-09-15T02:09:26+00:00  ip-10-0-49-55 !ERR     | scylla: [shard 7] view - Error applying view update to 10.0.141.40: exceptions::mutation_write_timeout_exception (Operation timed out for mview.users_by_first_name - received only 0 responses from 1 CL=ANY.)
```
job log:
scratch.scylladb.com/qa_issues/sct_15_54_26_09_2019.log.gz

**Comment by asias** [comment]
I looked at latest log from [above](https://github.com/scylladb/scylla/issues/4900#issuecomment-535573612)   scratch.scylladb.com/qa_issues/sct_15_54_26_09_2019.log.gz.

1) Repair stall is not the reason for the view update exception mutation_write_timeout_exception, because in 2019-09-15, there is zero stall on ip-10-0-49-55. [stall.txt](https://github.com/scylladb/scylla/files/3661220/stall.txt)


2) I decoded the stalls on ip-10-0-49-55 as well. [bt.txt](https://github.com/scylladb/scylla/files/3661202/bt.txt). None of them are related to repair.  


3) @psarna how long is the timeout for `view - Error applying view update to 10.0.141.40: exceptions::mutation_write_timeout_exception`

4) @slivne please assign to view experts. 

FYI: the binary is [here](http://downloads.scylladb.com.s3.amazonaws.com/downloads/scylla/rpm/centos/scylladb-3.1/7Server/x86_64/scylla-debuginfo-3.1.0.rc6-0.20190908.2ec036f50.el7.x86_64.rpm)

**Comment by psarna** [comment]
@asias 
Likely it's the famous 5 minute timeout:
~~~
storage_proxy.cc:1517
// View updates have a near-infinite timeout to avoid incurring the extra work of writting hints
// and to apply backpressure.
timeout = clock_type::now() + 5min;
~~~

**Comment by asias** [comment]
> @asias
> Likely it's the famous 5 minute timeout:
> 
> ```
> storage_proxy.cc:1517
> // View updates have a near-infinite timeout to avoid incurring the extra work of writting hints
> // and to apply backpressure.
> timeout = clock_type::now() + 5min;
> ```

OK. Thanks. This makes the stall-caused-view-update-timeout theory less likely. The biggest stall is 2.5s. It is impossible to make a 5 mins timeout to happen.

**Comment by yarongilor** [comment]
possibly reproduced in rc8 after stop-wait-start-scylla-service:
version: 3.1.0.rc8-0.20191001.57512d3df
AMI: ami-0c79c359ce06015ff
```
< t:2019-10-02 12:27:40,813 f:nemesis.py      l:338  c:sdcm.nemesis         p:DEBUG > sdcm.nemesis.ChaosMonkey: Set current_disruption -> StopWaitStartService Node longevity-tls-50gb-4d-3-1-db-node-4b0c3f59-2 [54.194.85.95 | 10.0.114.134] (seed: False)
< t:2019-10-02 12:33:54,497 f:nemesis.py      l:595  c:sdcm.nemesis         p:INFO  > sdcm.nemesis.ChaosMonkey: <<<<<<<<<<<<<Finished random_disrupt_method stop_wait_start_scylla_server
```

log: scratch.scylladb.com/qa_issues/sct_13_04_06_10_2019.log.gz

**Comment by slivne** [comment]
this is probably related to the issue we had in mv updates 2ff26d11608f26a129d3cdf2e173f8d39ae4b393

closing this issue - as I expect it to be resolved - if this happens again post 3.2.rc3 - please open a new issue
